nutils/sample.py

```diff
@@ -235,22 +234,14 @@ def __rmatmul__(self, func: function.Array) -> function.Array:
       return NotImplemented
     return _AtSample(func, self)
 
-  @property
-  def allcoords(self):
-    coords = numpy.empty([self.npoints, self.ndims])
-    for ielem, points in enumerate(self.points):
-      coords[self.getindex(ielem)] = points.coords
-    return types.frozenarray(coords, copy=False)
-
   def basis(self):
     '''Basis-like function that for every point in the sample evaluates to the
     unit vector corresponding to its index.'''
 
     index = function.transforms_index(self.transforms[0])
     coords = function.transforms_coords(self.transforms[0], self.ndims)
-    I = function.Elemwise(self.index, index, dtype=int)
-    B = function.Sampled(coords, expect=function.take(self.allcoords, I, axis=0))
-    return function.inflate(B, indices=I, length=self.npoints, axis=0)
+    sampled = function.Sampled(coords, expect=_Coords(self.points, index))
+    return function.inflate(sampled, indices=_Indices(self, index), length=self.npoints, axis=0)
```

Comment:
Instead of using `function.inflate`, you can make a single `function.Array` subclass that lowers to an `evaluable.Inflate`.

Reply:
done